src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas2/ApplicationNoteServiceTest.kt

```diff
@@ -0,0 +1,157 @@
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.Cas2ApplicationRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.AuthorisableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.results.ValidatableActionResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.NomisUserService
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.util.randomDateTimeBefore
+import java.time.OffsetDateTime
+import java.util.UUID
+
+class ApplicationNoteServiceTest {
+  private val mockApplicationRepository = mockk<Cas2ApplicationRepository>()
+  private val mockApplicationNoteRepository = mockk<Cas2ApplicationNoteRepository>()
+  private val mockUserService = mockk<NomisUserService>()
+
+  private val applicationNoteService = uk.gov.justice.digital.hmpps.approvedpremisesapi.service.cas2.ApplicationNoteService(
+    mockApplicationRepository,
+    mockApplicationNoteRepository,
+    mockUserService,
+  )
+
+  @Nested
+  inner class CreateApplicationNote {
+
+    private val referrer = NomisUserEntityFactory().produce()
+
+    @Nested
+    inner class WhenSuccessful {
+      private val submittedApplication = Cas2ApplicationEntityFactory()
+        .withCreatedByUser(referrer)
+        .withCrn("CRN123")
+        .withNomsNumber("NOMSABC")
+        .withSubmittedAt(OffsetDateTime.now().randomDateTimeBefore(2))
+        .produce()
+      private val applicationId = submittedApplication.id
+      private val noteEntity = Cas2ApplicationNoteEntity(
+        id = UUID.randomUUID(),
+        createdByNomisUser = referrer,
+        application = submittedApplication,
+        body = "new note",
+        createdAt = OffsetDateTime.now().randomDateTimeBefore(1),
+      )
+
+      @Test
+      fun `returns Success result with entity from db`() {
+        every { mockApplicationRepository.findByIdOrNull(applicationId) } answers
+          {
+            submittedApplication
+          }
+        every { mockUserService.getUserForRequest() } returns referrer
+        every { mockApplicationNoteRepository.save(any()) } answers
+          {
+            noteEntity
+          }
+
+        val result = applicationNoteService.createApplicationNote(
+          applicationId = applicationId,
+          NewCas2ApplicationNote(note = "new note"),
+        )
+
+        verify(exactly = 1) { mockApplicationNoteRepository.save(any()) }
+
+        Assertions.assertThat(result is AuthorisableActionResult.Success).isTrue
+        result as AuthorisableActionResult.Success
+
+        Assertions.assertThat(result.entity is ValidatableActionResult.Success).isTrue
+        val validatableActionResult = result.entity as ValidatableActionResult.Success
+
+        val createdNote = validatableActionResult.entity
+
+        Assertions.assertThat(createdNote).isEqualTo(noteEntity)
+      }
+    }
+
+    @Nested
+    inner class WhenUnsuccessful {
+      @Test
+      fun `returns Not Found when application not found`() {
```

Comment:
I would imagine there is a lot of duplicated code for some of these tests, they must be quite common across all the CAS2 use cases.